dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/map-to-fee-record-payment-groups.test.ts

```diff
@@ -0,0 +1,275 @@
+import { TfmUsersRepo } from '../../../../../repositories/tfm-users-repo';
+
+describe('mapToFeeRecordPaymentGroups', () => {
+  const findTfmUserSpy = jest.spyOn(TfmUsersRepo, 'findOneUserById');
+  const getFeeRecordPaymentEntityGroupStatusSpy = jest.spyOn(helpersModule, 'getFeeRecordPaymentEntityGroupStatus');
+  const getFeeRecordPaymentEntityGroupReconciliationDataSpy = jest.spyOn(helpersModule, 'getFeeRecordPaymentEntityGroupReconciliationData');
+
+  beforeEach(() => {
+    findTfmUserSpy.mockRejectedValue('Some error');
+    getFeeRecordPaymentEntityGroupStatusSpy.mockReturnValue(FEE_RECORD_STATUS.TO_DO);
+    getFeeRecordPaymentEntityGroupReconciliationDataSpy.mockResolvedValue({});
+  });
+
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  it('throws an error when a group has multiple fee records but no payments', async () => {
+    // Arrange
+    const group: FeeRecordPaymentEntityGroup = {
+      feeRecords: [
+        FeeRecordEntityMockBuilder.forReport(utilisationReport()).withId(1).build(),
+        FeeRecordEntityMockBuilder.forReport(utilisationReport()).withId(2).build(),
+        FeeRecordEntityMockBuilder.forReport(utilisationReport()).withId(3).build(),
+      ],
+      payments: [],
+    };
+
+    // Act / Assert
+    await expect(mapToFeeRecordPaymentGroups([group])).rejects.toThrow(
+      new Error('Fee record payment entity group cannot have more than one fee record if there are no payments'),
+    );
+  });
+
+  it('populates each of the groups with their respective group status', async () => {
+    // Arrange
+    const firstGroup: FeeRecordPaymentEntityGroup = {
+      feeRecords: [FeeRecordEntityMockBuilder.forReport(utilisationReport()).withStatus(FEE_RECORD_STATUS.TO_DO).build()],
+      payments: [],
+    };
+    const secondGroup: FeeRecordPaymentEntityGroup = {
+      feeRecords: [FeeRecordEntityMockBuilder.forReport(utilisationReport()).withStatus(FEE_RECORD_STATUS.DOES_NOT_MATCH).build()],
+      payments: [],
+    };
+    const thirdGroup: FeeRecordPaymentEntityGroup = {
+      feeRecords: [FeeRecordEntityMockBuilder.forReport(utilisationReport()).withStatus(FEE_RECORD_STATUS.READY_TO_KEY).build()],
+      payments: [],
+    };
+    const groups = [firstGroup, secondGroup, thirdGroup];
+
+    when(getFeeRecordPaymentEntityGroupStatusSpy).calledWith(firstGroup).mockReturnValue(FEE_RECORD_STATUS.TO_DO);
+    when(getFeeRecordPaymentEntityGroupStatusSpy).calledWith(secondGroup).mockReturnValue(FEE_RECORD_STATUS.DOES_NOT_MATCH);
+    when(getFeeRecordPaymentEntityGroupStatusSpy).calledWith(thirdGroup).mockReturnValue(FEE_RECORD_STATUS.READY_TO_KEY);
+
+    // Act
+    const result = await mapToFeeRecordPaymentGroups(groups);
+
+    // Assert
+    expect(result).toHaveLength(3);
+    expect(result[0].status).toBe(FEE_RECORD_STATUS.TO_DO);
+    expect(result[1].status).toBe(FEE_RECORD_STATUS.DOES_NOT_MATCH);
+    expect(result[2].status).toBe(FEE_RECORD_STATUS.READY_TO_KEY);
```

Comment:
Why are we testing statuses here there's nothing about this directly in `mapToFeeRecordPaymentGroups`, rather it calls a series of other functions.

E.g, these functions are called but there's nothing for these in the unit tests?

```ts
calculateTotalCurrencyAndAmount
mapPaymentEntityToPayment
calculateTotalCurrencyAndAmount
```

Reply:
We indirectly ensure that the functions are called via `jest-when`. In this case, we ensure that the specific call of `getFeeRecordPaymentEntityGroupStatus` returns the correct status, and that the ordering of this status is maintained in the returned groups